index/scorch/optimize_knn.go

```diff
@@ -36,35 +37,48 @@ func (o *OptimizeVR) Finish() error {
 	// for each VR, populate postings list and iterators
 	// by passing the obtained vector index and getting similar vectors.
 	// defer close index - just once.
-
+	errors := make([]error, len(o.snapshot.segment))
+	wg := sync.WaitGroup{}
+	// Launch goroutines to get vector index for each segment
 	for i, seg := range o.snapshot.segment {
-		// for each field, get the vector index --> invoke the zap func.
-		for field, vrs := range o.vrs {
-			// for each VR belonging to that field
-			if sv, ok := seg.segment.(segment_api.VectorSegment); ok {
-				// reading just once per field per segment.
-				searchVectorIndex, closeVectorIndex, err := sv.InterpretVectorIndex(field)
-				if err != nil {
-					return err
-				}
-
-				for _, vr := range vrs {
-					// for each VR, populate postings list and iterators
-					// by passing the obtained vector index and getting similar vectors.
-					pl, err := searchVectorIndex(vr.field, vr.vector, vr.k, seg.deleted)
+		if sv, ok := seg.segment.(segment_api.VectorSegment); ok {
+			wg.Add(1)
+			go func(index int, segment segment_api.VectorSegment, origSeg *SegmentSnapshot) {
+				defer wg.Done()
+				for field, vrs := range o.vrs {
+					searchVectorIndex, closeVectorIndex, err := segment.InterpretVectorIndex(field)
+					defer func() {
+						if closeVectorIndex != nil {
```

Comment:
Remove this nil check and move this defer to 58.5 below, because if err == nil, neither searchVectorIndex nor closeVectorIndex will ever be nil.